core/access/chain_access.go

```diff
@@ -0,0 +1,219 @@
+}
+
+// OdrEnabled returns true if this ChainAccess is capable of doing ODR requests
+func (self *ChainAccess) OdrEnabled() bool {
+	return self.odr
+}
+
+// RegisterPeer registers a new LES peer to the ODR capable peer set
+func (self *ChainAccess) RegisterPeer(id string, version int, head common.Hash, getBlockBodies getBlockBodiesFn, getNodeData getNodeDataFn, getReceipts getReceiptsFn, getProofs getProofsFn) error {
+	glog.V(logger.Detail).Infoln("Registering peer", id)
+	if err := self.peers.Register(newPeer(id, version, head, getBlockBodies, getNodeData, getReceipts, getProofs)); err != nil {
+		glog.V(logger.Error).Infoln("Register failed:", err)
+		return err
+	}
+	return nil
+}
+
+// UnregisterPeer removes a peer from the ODR capable peer set
+func (self *ChainAccess) UnregisterPeer(id string) {
+	self.peers.Unregister(id)
+}
+
+const (
+	MsgBlockBodies = iota
+	MsgNodeData
+	MsgReceipts
+	MsgProofs
+)
+
+// Msg encodes a LES message that delivers reply data for a request
+type Msg struct {
+	MsgType int
+	Obj     interface{}
+}
+
+// ObjectAccess is the ODR request interface (passed to Retrieve, functions called by Retrieve and Deliver)
+// 		DbGet() tries to retrieve the object from the local database (object is stored by the request struct in memory if retrieved)
+//		DbPut() stores it in the local database
+//		Request(*Peer) requests it from a LES peer
+//		Valid(*Msg) checks if a message is a valid answer to this request and stores the retrieved object in memory
+type ObjectAccess interface {
+	// database storage
+	DbGet() bool
+	DbPut()
+	// network retrieval
+	Request(*Peer) error
+	Valid(*Msg) bool // if true, keeps the retrieved object
+}
+
+// Deliver is called by the LES protocol manager to deliver ODR reply messages to waiting requests
+func (self *ChainAccess) Deliver(id string, msg *Msg) (processed bool) {
+	self.lock.Lock()
+	defer self.lock.Unlock()
+
+	for i, req := range self.sentReqs {
+		if req.valFunc(msg) {
+			req.deliverChan <- msg
+			delete(self.sentReqs, i)
+			return true
+		}
+	}
+	return false
+}
+
+// networkRequest sends a request to known peers until an answer is received
+// or the context is cancelled
+func (self *ChainAccess) networkRequest(ctx context.Context, rqFunc requestFunc, valFunc validatorFunc) (*Msg, error) {
+	req := &sentReq{
+		deliverChan: make(chan *Msg),
+		valFunc:     valFunc,
+	}
+	self.lock.Lock()
+	reqCnt := self.sentReqCnt
+	self.sentReqCnt++
+	self.sentReqs[reqCnt] = req
+	self.lock.Unlock()
+
+	defer func() {
+		self.lock.Lock()
+		delete(self.sentReqs, reqCnt)
+		self.lock.Unlock()
+	}()
+
+	var msg *Msg
+
+	for {
+		peers := self.peers.BestPeers()
+		if len(peers) == 0 {
+			select {
+			case <-ctx.Done():
+				setTerminated(ctx)
+				return nil, ctx.Err()
+			case <-time.After(retryPeers):
+			}
+		}
+		for _, peer := range peers {
+			rqFunc(peer)
+			select {
+			case <-ctx.Done():
+				setTerminated(ctx)
+				return nil, ctx.Err()
+			case msg = <-req.deliverChan:
```

Comment:
AFAIU this is problematic, since an earlier peer can also deliver here (delivery is not syncronised with the timeout here.) on line 188 you reward another peer for their late message.
also line 182 request may typically block on other network writes or slow connections, so may want to do writesin a channel skipping busy peers.
in general you probably want to ask several peers parallel, in this case however you cannot demote peers that time out, only promote the winner. in order to normalise promotion, might make sense to ask fixed number of peers (3)